File: src/core/paths.ts
/**
 * Locating session files on disk for each agent, cross-platform.
 *
 * Claude:  <claudeRoot>/projects/<encoded-cwd>/<session-uuid>.jsonl
 *          where <encoded-cwd> is the absolute cwd with every run of non
 *          [A-Za-z0-9] characters replaced by '-'.
 * Codex:   <codexRoot>/sessions/YYYY/MM/DD/rollout-<ts>-<session-uuid>.jsonl
 *          (cwd lives inside the file's session_meta, not the path).
 *
 * Multiple roots are supported so xfer can index sessions from the CLI install,
 * a desktop app data dir, and third-party/gateway-mode installs at once. Extra
 * roots come from XFER_CLAUDE_EXTRA_ROOTS / XFER_CODEX_EXTRA_ROOTS, each an OS
 * path-list (a;b on Windows, a:b on POSIX).
 */

import os from "node:os";
import path from "node:path";
import fs from "node:fs";

export function claudeHome(): string {
  // Claude Code reads CLAUDE_CONFIG_DIR for an alternate config/data location.
  return (
    process.env.CLAUDE_CONFIG_DIR ||
    process.env.CLAUDE_HOME ||
    path.join(os.homedir(), ".claude")
  );
}

export function codexHome(): string {
  return process.env.CODEX_HOME || path.join(os.homedir(), ".codex");
}

/** Split an OS path-list env value into trimmed, non-empty roots. */
export function splitRoots(value: string | undefined): string[] {
  if (!value) return [];
  return value
    .split(path.delimiter)
    .map((s) => s.trim())
    .filter(Boolean);
}

/** All Claude roots: the primary home plus any extra configured roots. */
export function claudeRoots(): string[] {
  return [claudeHome(), ...splitRoots(process.env.XFER_CLAUDE_EXTRA_ROOTS)];
}

/** All Codex roots: the primary home plus any extra configured roots. */
export function codexRoots(): string[] {
  return [codexHome(), ...splitRoots(process.env.XFER_CODEX_EXTRA_ROOTS)];
}

/** Primary Claude projects dir (used for writing migrated Claude sessions). */
export function claudeProjectsDir(): string {
  return path.join(claudeHome(), "projects");
}

/** Primary Codex sessions dir (used for writing migrated Codex sessions). */
export function codexSessionsDir(): string {
  return path.join(codexHome(), "sessions");
}

/** Every Claude `projects` dir across all roots (for read/index scanning). */
export function claudeProjectDirs(): string[] {
  return claudeRoots().map((r) => path.join(r, "projects"));
}

/** Every Codex `sessions` dir across all roots (for read/index scanning). */
export function codexSessionDirs(): string[] {
  return codexRoots().map((r) => path.join(r, "sessions"));
}

export function normalizeProjectCwd(cwd: string): string {
  if (cwd.startsWith("\\\\?\\UNC\\")) return `\\\\${cwd.slice("\\\\?\\UNC\\".length)}`;
  if (/^\\\\\?\\[A-Za-z]:[\\/]/.test(cwd)) return cwd.slice(4);
  return cwd;
}

/**
 * Encode an absolute cwd the way Claude Code names its project folders.
 * Observed rule: replace every character that is not a letter or digit with '-'.
 * e.g. `E:\0001_Work\foo` -> `E--0001-Work-foo`.
 */
export function encodeClaudeProjectDir(cwd: string): string {
  return normalizeProjectCwd(cwd).replace(/[^A-Za-z0-9]/g, "-");
}

/** Recursively collect all *.jsonl files under a directory (returns [] if missing). */
export function walkJsonl(dir: string): string[] {
  const out: string[] = [];
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return out;
  }
  for (const e of entries) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) out.push(...walkJsonl(p));
    else if (e.isFile() && e.name.endsWith(".jsonl")) out.push(p);
  }
  return out;
}

/** Extract a UUID embedded anywhere in a string (session id from file names). */
export function extractUuid(s: string): string | undefined {
  const m = s.match(
    /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/i
  );
  return m ? m[0] : undefined;
}

File: src-tauri/src/xfer/paths.rs
use std::path::{Path, PathBuf};
use walkdir::WalkDir;

pub fn home_dir() -> PathBuf {
  dirs::home_dir().unwrap_or_else(|| std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")))
}

pub fn claude_home() -> PathBuf {
  std::env::var_os("CLAUDE_CONFIG_DIR")
    .or_else(|| std::env::var_os("CLAUDE_HOME"))
    .map(PathBuf::from)
    .unwrap_or_else(|| home_dir().join(".claude"))
}

pub fn codex_home() -> PathBuf {
  std::env::var_os("CODEX_HOME")
    .map(PathBuf::from)
    .unwrap_or_else(|| home_dir().join(".codex"))
}

/// Split an OS path-list env value into trimmed, non-empty roots.
pub fn split_roots(value: Option<std::ffi::OsString>) -> Vec<PathBuf> {
  let Some(value) = value else { return Vec::new() };
  std::env::split_paths(&value).filter(|p| !p.as_os_str().is_empty()).collect()
}

/// All Claude roots: the primary home plus any extra configured roots.
pub fn claude_roots() -> Vec<PathBuf> {
  let mut roots = vec![claude_home()];
  roots.extend(split_roots(std::env::var_os("XFER_CLAUDE_EXTRA_ROOTS")));
  roots
}

/// All Codex roots: the primary home plus any extra configured roots.
pub fn codex_roots() -> Vec<PathBuf> {
  let mut roots = vec![codex_home()];
  roots.extend(split_roots(std::env::var_os("XFER_CODEX_EXTRA_ROOTS")));
  roots
}

pub fn claude_projects_dir() -> PathBuf {
  claude_home().join("projects")
}

pub fn codex_sessions_dir() -> PathBuf {
  codex_home().join("sessions")
}

/// Every Claude `projects` dir across all roots (for read/index scanning).
pub fn claude_project_dirs() -> Vec<PathBuf> {
  claude_roots().into_iter().map(|r| r.join("projects")).collect()
}

/// Every Codex `sessions` dir across all roots (for read/index scanning).
pub fn codex_session_dirs() -> Vec<PathBuf> {
  codex_roots().into_iter().map(|r| r.join("sessions")).collect()
}

pub fn encode_claude_project_dir(cwd: &str) -> String {
  normalize_project_cwd(cwd)
    .chars()
    .map(|c| if c.is_ascii_alphanumeric() { c } else { '-' })
    .collect()
}

pub fn normalize_project_cwd(cwd: &str) -> String {
  strip_windows_verbatim_prefix(cwd)
}

pub fn strip_windows_verbatim_prefix(path: &str) -> String {
  if let Some(rest) = path.strip_prefix(r"\\?\UNC\") {
    return format!(r"\\{rest}");
  }
  if let Some(rest) = path.strip_prefix(r"\\?\") {
    return rest.to_string();
  }
  path.to_string()
}

pub fn extract_uuid(input: &str) -> Option<String> {
  let bytes = input.as_bytes();
  for i in 0..bytes.len().saturating_sub(35) {
    let s = &input[i..i + 36];
    if is_uuid_like(s) {
      return Some(s.to_string());
    }
  }
  None
}

fn is_uuid_like(s: &str) -> bool {
  let bs = s.as_bytes();
  if bs.len() != 36 {
    return false;
  }
  for (i, b) in bs.iter().enumerate() {
    if matches!(i, 8 | 13 | 18 | 23) {
      if *b != b'-' {
        return false;
      }
    } else if !b.is_ascii_hexdigit() {
      return false;
    }
  }
  true
}

pub fn walk_jsonl(dir: &Path) -> Vec<PathBuf> {
  if !dir.exists() {
    return Vec::new();
  }
  WalkDir::new(dir)
    .into_iter()
    .filter_map(Result::ok)
    .filter(|entry| entry.file_type().is_file())
    .map(|entry| entry.into_path())
    .filter(|path| path.extension().and_then(|s| s.to_str()) == Some("jsonl"))
    .collect()
}

pub fn normalize_cwd(cwd: &str) -> String {
  normalize_project_cwd(cwd).replace('\\', "/").trim_end_matches('/').to_lowercase()
}

pub fn default_project_cwd(cwd: Option<&str>) -> String {
  let path = cwd.map(PathBuf::from).unwrap_or_else(|| std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")));
  normalize_project_cwd(&path.canonicalize().unwrap_or(path).to_string_lossy())
}
